webcrypto: Reduce usage of standalone helper functions for JWK format

kkoyung opened the pull request:
Reduce the reliance on standalone helper functions for handling JWK format. Instead, those functionalities are now integrated into the `JsonWebKey` type generated by script_binding, via the local trait `JsonWebKeyExt`, for internal use.

The `parse_jwk` function remains for now. It will be removed when once we refactor `SubtleCrypto::ImportKey` to support a more generic approach across different cryptographic algorithms.

Testing: Refactoring. Existing WPT tests should suffice.

kkoyung committed:
Extend functionalities of RsaOtherPrimesInfo and JsonWebKey

Add extra methods for those types from script-binding for internal use.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Replace usage_from_str by KeyUsage::from_str

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Replace standalone JsonWebKey helper functions

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Resolve clippy warning

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

jdm commented on components/script/dom/subtlecrypto.rs:
```diff
@@ -3203,3 +3204,158 @@ fn usage_from_str(op: &str) -> Result<KeyUsage, Error> {
+        // Step 4. Let result be the object literal that results from executing the JSON.parse
+        // internal function in the context of a new global object, with text argument set to a
+        // JavaScript String containing json.
+        let value = serde_json::from_slice(data)
```
Rather than handwriting all of this, I think we should follow the spec and do:
```rust
let data = str::from_utf8(data).map_err(|_| Error::Data)?;
let data: Vec<_> = data.encode_utf16().collect();
rooted!(in(*cx) let mut json_value = UndefinedValue());
if !JS_ParseJSON(cx, data.as_ptr(), data.len(), json_value.handle_mut())) {
    return Error::JSFailed;
}
```
We can then do step 5 as: `JsonWebKey::new(cx, json_value.handle())`.

kkoyung replied:
Great recommendation. It also makes the code shorter.

jdm reviewed (changes requested):
Looks like a good cleanup, but we can go deeper!

kkoyung committed:
Further follow the spec

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Missing unused annotation

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

kkoyung committed:
Add incompatible_msrv annotation

`str::from_utf8` is stable starting from Rust 1.87.0, while current MSRV
(Minimum Supported Rust Version) is 1.85.0.

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

jdm commented on components/script/dom/subtlecrypto.rs:
```diff
@@ -3202,6 +3202,7 @@ impl JsonWebKeyExt for JsonWebKey {
         // (It is given as a method paramter.)
 
         // Step 2. Let json be the Unicode string that results from interpreting data according to UTF-8.
+        #[allow(clippy::incompatible_msrv)]
```
Ah, this will break when we try to merge on CI, since we check that out MSRV is accurate. Let's use https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy instead.

kkoyung committed:
Use String::from_utf8_lossy instead of str::from_utf8

Signed-off-by: Kingsley Yung <kingsley@kkoyung.dev>

jdm reviewed (approved):
Thanks!
